### PR TITLE
autocast: force CUDA FP16 to avoid BF16 instability

### LINKED ISSUE #173066: Training accuracy may suddenly fall to near 0 in the early training stage using pytorch 2.9 and AMP (bfloat16)
### 🐛 Describe the bug

I am training a deep convolutional neural network for a custom image classification task using ConvNext-v2 backbone. In the early training stage, for example in warmup or the first 80 epochs, the training accuracy may suddenly fall to near 0 and then recover to the original accuracy in a few epochs. I use AdamW optimizer with AMP (bfloat16), and learning rate scheduling with linear warmup and Cosine annealing, and gradients accumulation to accelerate training on a H20 gpu. The batch size is 512, and the accumulated batch size is 4096. The pytorch and CUDA version are 2.9.0 and 12.8, respectively. I have tried to extend the warmup epochs or decrease the learning rate but failed.

Curiously if I add the wandb watch function in the training loop, this phenomenon will not occur.

However, this phenomenon has never occurred before when running the same code and the same training strategy with batchsize=128 and accumulated  batchsize=1024, and pytorch+CUDA=2.6.0+12.4 on a 3090 GPU.

What are the possible reasons?

### Versions

Pytorch 2.9.0
CUDA 12.8
Python 3.12

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by VengadeshwararRao**
This PR forces FP16 under CUDA autocast to avoid BF16-related instability.

Context:
- Observed incorrect BF16 selection in CUDA autocast path
- Causes runtime instability / incorrect behavior in affected setups

Change:
- Explicitly enforces FP16 for CUDA autocast

No functional impact outside CUDA autocast.
Fixes #173066

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: aten/src/ATen/autocast_mode.cpp [hunk 1]
```diff
@@ -100,9 +100,13 @@ int decrement_nesting() {
 }
 
 at::ScalarType get_autocast_dtype(at::DeviceType device_type) {
+  if (device_type == at::DeviceType::CUDA) {
+    return at::kHalf;  // FORCE FP16, ignore BF16
```
Comment (guangyey) [inline aten/src/ATen/autocast_mode.cpp:104]: Personally, I don't think it is reasonable to add this here. Request the final stamp from Meta and CUDA team.

**Review by eqy — CHANGES_REQUESTED** [review]
Even if we were to change a default, we would not do it by hardcoding it here and potentially confusing users

There are many tradeoffs between bf16 vs. fp16 stability (consider dynamic range) so we would not merge something like this
